### PR TITLE
Git2941 Owl RDF Canonical glossary parser  

### PR DESCRIPTION
**Pull request by davidradl**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -89,12 +89,10 @@ egeria-lineage-repositories/**
 *-graph-repository/**
 
 # Ignore archives
-CloudInformationModel.json
-CloudInformationModelGUIDMap.json
-DataStoreConnectorTypes.json
-DataStoreConnectorsGUIDMap.json
-OpenMetadataTypes.json
+*.json
```
Comment (planetf1) [inline .gitignore:92]: I wonder if this exclude is rather too general -- will we not have .json files in future?

If the files are GENERATED they should be under the target tree - those files are already entirely excluded from git.

See below for unit test too - we should generate them into a target directory

This helps keep excludes under control, and also makes it easier for packaging and assembly

**Comment by planetf1** [comment]
In #2788 I tried to fix the creation -- not completed as I hit an issue -- of the archive for CIM, and the fix is bad hence closed, but will revisit under #2778 in the 1.8 timescale (also changes needed to other clients) and so this could be good to cover also.

File: .gitignore (hunk 1 above)
Comment (planetf1) [reply .gitignore:92]: Another reason for this is that other build lifecycle aspects like 'clean' will work properly without extra work, as well as making it easier to run other utilities across the source - for docs etc.

Unfortunately in both the UI and the subject area tests this wasn't done which caused a fair bit of heartache to fix up some consequential issues.

Comment (davidradl) [reply .gitignore:92]: @planetf1 I agree normally. I have changed the logic so the json file is generated from the model name - so we do not know it advance. It is the existing logic for the archive writer to write the output here - I have not changed that. As far as my change -there is a json file which is the archive which is picked up from here; this is the same as a the CIM model. There is a guid file is generated (also existing logic) - but you would not necessarily want it deleted as part of a maven clean - as it is the file that ensures subsequent builds use the same guids.    I suggest we raise a separate issue to review the input and output files locations as I see this as a separate issue; we can amend the .gitignore at that stage if required. Is that ok with you? 
I will look at the junit as that should write to target.

**Comment by davidradl** [comment]
@planetf1  I have amended the code so the junit does not write to the file system (googling around this seems like bad practice). I have changed the gitignore to be more targetted.

**Comment by planetf1** [comment]
> @planetf1 I have amended the code so the junit does not write to the file system (googling around this seems like bad practice). I have changed the gitignore to be more targetted.

Thanks. even better if we don't need to write.

I notice in the gitignore we have:
```
*Archive.json
*GUIDMap.json

# Ignore any json files generated during junit
open-metadata-resources/open-metadata-archives/design-model-archives/*.json
```
Where do these get written to (I haven't had a chanbce to try it)

They should only get written to the target directory. Adding anything created elsewhere is bad practice, and can cause problems with packaging.

Part of this is that we need to ensure that 'mvn clean' restores the files to the same state beforehand.. but there are other implications ie on wildcards for packaging. Best practice is to output all files into 'target' only.

If that's the case no gitignores would be needed.

We do breach this slightly on the top level with the config/registry files (there is an issue open to address this)

**Comment by davidradl** [comment]
@planetf1. The reason for the need for the .gitignore is not to ignore files that are created during the build. The reason is that, in a dev environment in IJ, when you run the archive parser writer builder, then it looks for the the guid file in the root folder, it generates the archive where the model file is. I then copy the archive file into the root folder. I then start the omag server. I then issue the rest call to load the archive - the filename being in the rest body. This is how I get Egeria to pick up the archive. I have not changed this mechanism, this was the same for cim, but it had hard coded names.

So the .gitignore is required to ensure that in a dev environment these files are not inadvertently put into git. I suggest any enhancement to this existing mechanism be pursued in a separate issue.

**Comment by planetf1** [comment]
Agreed - thanks for the explanation. JUst wanted to check I wasn't missing anything done in the build -- which as you say we'll tackle in the related issue.

**Comment by davidradl** [comment]
I intend to merge this at the beginning of April the 29th, unless I receive more feedback

File: open-metadata-implementation/repository-services/repository-services-archive-utilities/src/main/java/org/odpi/openmetadata/repositoryservices/archiveutilities/OMRSArchiveGUIDMap.java [hunk 2]
```diff
@@ -88,8 +88,21 @@ public String  getGUID(String id)
 
         if (guid == null)
         {
-            guid = UUID.randomUUID().toString();
-
+            boolean duplicate =true;
```
Comment (mandy-chessell) [inline open-metadata-implementation/repository-services/repository-services-archive-utilities/src/main/java/org/odpi/openmetadata/repositoryservices/archiveutilities/OMRSArchiveGUIDMap.java]: What is it likelihood of getting a GUID clash from RandomUUID given it uses a timestamp?

File: open-metadata-implementation/user-interfaces/ui-chassis/ui-chassis-spring/src/main/java/org/odpi/openmetadata/userinterface/uichassis/springboot/api/rex/RepositoryExplorerController.java [hunk 3]
```diff
@@ -496,7 +496,7 @@ public RexTraversalResponse rexTraversal(@RequestBody RexTraversalRequestBody bo
         try {
             serverName          = body.getServerName();
             serverURLRoot       = body.getServerURLRoot();
-            enterpriseOption    = body.getEnterpriseOption();
+             enterpriseOption    = body.getEnterpriseOption();
```
Comment (mandy-chessell) [inline open-metadata-implementation/user-interfaces/ui-chassis/ui-chassis-spring/src/main/java/org/odpi/openmetadata/userinterface/uichassis/springboot/api/rex/RepositoryExplorerController.java]: Fix alignment

File: open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/base/DesignModelArchiveBuilder.java [hunk 4]
```diff
@@ -295,18 +326,65 @@ protected String addTerm(String       glossaryId,
                              List<String> categoryIds,
                              String       qualifiedName,
                              String       displayName,
-                             String       description)
+                             String       description) {
```
Comment (mandy-chessell) [inline open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/base/DesignModelArchiveBuilder.java:330]: Formatting

File: open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/base/DesignModelArchiveBuilder.java [hunk 5]
```diff
@@ -326,6 +404,9 @@ protected String addTerm(String       glossaryId,
             {
                 if (categoryId != null)
                 {
+                    if (categoriesAsNames) {
```
Comment (mandy-chessell) [inline open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/base/DesignModelArchiveBuilder.java:409]: Formatting

File: open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/base/DesignModelArchiveBuilder.java [hunk 6]
```diff
@@ -754,6 +835,70 @@ protected void addLinkToBead(String                 beadId,
...
+
+        archiveBuilder.addRelationship(archiveHelper.getRelationship(ISA_RELATIONSHIP_NAME,
+                                                                     idToGUIDMap.getGUID(generalizededTermId + "_to_" + specializedTermId),
+                                                                     null,
+                                                                     InstanceStatus.ACTIVE,
+                                                                     end1,
+                                                                     end2));
+    }
+    protected void addHasARelationship(String conceptQName, String propertyQName) {
+        String conceptId = idToGUIDMap.getGUID(conceptQName);
```
Comment (mandy-chessell) [inline open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/base/DesignModelArchiveBuilder.java:862]: Formatting - problem continues futher down file

File: open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/owlcanonicalglossarymodel/CanonicalGlossaryOwlArchiveWriter.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+    private String modelLocation;
+
+
+    /**
+     * Default constructor
+     *
+     * @param modelLocation directory name for the model's JSON-LD files.
+     */
+    CanonicalGlossaryOwlArchiveWriter(String modelLocation)
```
Comment (mandy-chessell) [inline open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/owlcanonicalglossarymodel/CanonicalGlossaryOwlArchiveWriter.java:25]: Should this name be more specific as only works for industry models ?

File: open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/owlcanonicalglossarymodel/properties/ModelElement.java [hunk 8]
```diff
@@ -0,0 +1,226 @@
+/* SPDX-License-Identifier: Apache 2.0 */
+/* Copyright Contributors to the ODPi Egeria project. */
+package org.odpi.openmetadata.archiveutilities.designmodels.owlcanonicalglossarymodel.properties;
+
+import java.util.*;
+
+/**
+ * Attribute describes an attribute in the model
+ */
+public class ModelElement
```
Comment (mandy-chessell) [inline open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/owlcanonicalglossarymodel/properties/ModelElement.java:10]: Javadoc missing on methods - this is a generic comment as occurs in other files

File: open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/test/JenaTest.java [hunk 9]
```diff
@@ -0,0 +1,549 @@
+/* SPDX-License-Identifier: Apache 2.0 */
```
Comment (mandy-chessell) [inline open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/test/JenaTest.java]: Seems this is in the wrong package - should be in the OWL package

Comment (davidradl) [reply open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/test/JenaTest.java]: I think you are looking at an old version of the fix this file is not in the current fix

File: open-metadata-implementation/repository-services/repository-services-archive-utilities/src/main/java/org/odpi/openmetadata/repositoryservices/archiveutilities/OMRSArchiveGUIDMap.java (hunk 2 above)
Comment (davidradl) [reply open-metadata-implementation/repository-services/repository-services-archive-utilities/src/main/java/org/odpi/openmetadata/repositoryservices/archiveutilities/OMRSArchiveGUIDMap.java]: I think you are looking at an old version of the fix this code is not in the current fix

File: open-metadata-implementation/user-interfaces/ui-chassis/ui-chassis-spring/src/main/java/org/odpi/openmetadata/userinterface/uichassis/springboot/api/rex/RepositoryExplorerController.java (hunk 3 above)
Comment (davidradl) [reply open-metadata-implementation/user-interfaces/ui-chassis/ui-chassis-spring/src/main/java/org/odpi/openmetadata/userinterface/uichassis/springboot/api/rex/RepositoryExplorerController.java]: I think you are looking at an old version of the fix this file is not in the current fix

File: open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/base/DesignModelArchiveBuilder.java (hunk 4 above)
Comment (davidradl) [reply open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/base/DesignModelArchiveBuilder.java]: fixed

File: open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/base/DesignModelArchiveBuilder.java (hunk 6 above)
Comment (davidradl) [reply open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/base/DesignModelArchiveBuilder.java:863]: I think you are looking at a back level fix - this is not in the current fix

File: open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/owlcanonicalglossarymodel/CanonicalGlossaryOwlArchiveWriter.java (hunk 7 above)
Comment (davidradl) [reply open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/owlcanonicalglossarymodel/CanonicalGlossaryOwlArchiveWriter.java:25]: We hope that others in the community will adopt this way of modelling if it suits them, there is nothing in the format that is intended to be IBM specific.

File: open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/base/DesignModelArchiveBuilder.java (hunk 5 above)
Comment (davidradl) [reply open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/base/DesignModelArchiveBuilder.java]: fixed

File: open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/owlcanonicalglossarymodel/properties/ModelElement.java (hunk 8 above)
Comment (davidradl) [reply open-metadata-resources/open-metadata-archives/design-model-archives/src/main/java/org/odpi/openmetadata/archiveutilities/designmodels/owlcanonicalglossarymodel/properties/ModelElement.java]: fixed

**Pull request merged**